### PR TITLE
HZN-1398: Prevent cyclic graphs with situations

### PR DESCRIPTION
**Pull request by cgorantla**
* JIRA: https://issues.opennms.org/browse/HZN-1398

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

### CODE REVIEW

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java [hunk 1]
```diff
@@ -345,6 +347,23 @@ private OnmsAlarm createNewAlarm(OnmsEvent e, Event event) {
         e.setAlarm(alarm);
         return alarm;
     }
+
+
+    private boolean formingCyclicGraph(OnmsAlarm situation, OnmsAlarm relatedAlarm) {
```
[the remark this answers is not in the document]
Comment (j-white) [reply opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: At first glance, this doesn't appear to catch the case where a loop would be formed further down the tree. The model currently allows for situations to contain other situations, and so on.

[the remark this answers is not in the document]
Comment (mattixtech) [reply opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: Oops, you are right. I think this should fix that:

`private boolean formingCyclicGraph(OnmsAlarm situation, OnmsAlarm relatedAlarm) {
        return  situation.getReductionKey().equals(relatedAlarm.getReductionKey()) || relatedAlarm.getRelatedAlarms()
                .stream().anyMatch(ra -> formingCyclicGraph(situation, ra));
}`

[the remark this answers is not in the document]
Comment (cgorantla) [reply opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: yup, that should do it.  code reading wise streams one seems to be little better.

**Comment by mattixtech** [comment]
👍

**Pull request merged**